# The Server Lag (Normality Test & Data Cleaning)

### Objective:
Simulate 1,000 normal server response times, corrupt the dataset with 8 severe network spikes, and observe how the Shapiro-Wilk test reacts. Then, use Boolean masking to clean the data and restore normality.

>Test Used: Shapiro-Wilk Test of Normality

>Performed by : Raghav Katyal on 28 September 2026

In [532]:
#importing required libraries to perform the simulation and test
import numpy as np
from scipy import stats

In [533]:
#fixing the seed so that random numbers remain stable
np.random.seed(42)

In [534]:
#Creating simulated server load times (in ms)
#Average Load Time is 200ms with standard deviation of 20ms
#A Sample of 1000 size is taken
load_times = np.random.normal(loc=200,scale=20,size=1000)

In [535]:
#Shapiro-Wilk Test (Before adding corrupted timings)
#It should show perfectly normal (expected) W-statistic and p-value.
w_value,p_value = stats.shapiro(load_times)
print(f"The W-statistic before adding corrupted data is : {w_value:.4f}")
print(f"The p-value before adding corrupted data is : {p_value:.2f}")


The W-statistic before adding corrupted data is : 0.9986
The p-value before adding corrupted data is : 0.63


>Observation:
A W-statistic value of 0.99 and p-value of 0.63 gives no strong evidence against Normality ; hence data is considered normally distributed.

In [536]:
#corrupting the data with 500ms load timings by replacing last 8 values with 500ms
corrupt_load_times = load_times.copy()
corrupt_load_times[-8:] = 500

In [537]:
#Running Shapiro-Wilk test again
w_value,p_value = stats.shapiro(corrupt_load_times)
print(f"The W-statistic after adding corrupted data is : {w_value:.4f}")
print(f"The p-value after adding corrupted data is : {p_value:.4f}")

The W-statistic after adding corrupted data is : 0.5916
The p-value after adding corrupted data is : 0.0000


>Observation:
A W-statistic value of 0.59 and p-value of 0.00 aggressively shows evidence against Normality ; hence data is not at all considered normally distributed.

In [538]:
#Boolean Masking to Filter the Data and thus removing the aggressive load timings
cleaned_corrupt_load_times = corrupt_load_times[corrupt_load_times < 400]

#testing again
w_value,p_value=stats.shapiro(cleaned_corrupt_load_times)
print(f"The W-statistic after cleaning corrupted data is : {w_value:.4f}")
print(f"The p-value after cleaning corrupted data is : {p_value:.4f}")


The W-statistic after cleaning corrupted data is : 0.9987
The p-value after cleaning corrupted data is : 0.6710


>Observation:
A W-statistic value of 0.99 and p-value of 0.67 shows no strong evidence against Normality ; hence data is again considered normally distributed.